components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,251 @@
+ * @throws Propagates `update_offset`'s exceptions.
+ */
+[[nodiscard]] auto
+get_file_infos(std::filesystem::path const& multi_file_archive_path, segment_id_t next_segment_id)
+        -> std::vector<FileInfo>;
+
+/**
+ * Combines file section metadata and the number of segments into single-file archive metadata,
+ * then serializes the metadata into MsgPack format.
+ *
+ * @param multi_file_archive_path
+ * @param next_segment_id
+ * @return Packed metadata.
+ */
+[[nodiscard]] auto pack_single_file_archive_metadata(
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> std::stringstream;
+
+/**
+ * Writes single-file archive header.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata_size
+ */
+auto write_archive_header(FileWriter& single_file_archive_writer, size_t packed_metadata_size)
+        -> void;
+
+/**
+ * Writes single-file archive metadata.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata Packed metadata.
+ */
+auto write_archive_metadata(
+        FileWriter& single_file_archive_writer,
+        std::stringstream const& packed_metadata
+) -> void;
+
+/**
+ * Reads the content of a file and writes it to the single-file archive.
+ * @param file_path
+ * @param single_file_archive_writer
+ * @throws OperationFailed if reading the file fails.
+ */
+auto write_archive_file(std::string const& file_path, FileWriter& single_file_archive_writer)
+        -> void;
+
+/**
+ * Iterates over files in the multi-file archive and copies their contents to the single-file
+ * archive.
+ *
+ * @param single_file_archive_writer
+ * @param multi_file_archive_path
+ * @param next_segment_id
+ * @throws Propagates `update_offset`'s exceptions.
+ */
+auto write_archive_files(
+        FileWriter& single_file_archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void;
+
+auto get_file_size_and_update_offset(std::filesystem::path const& file_path, uint64_t& offset)
+        -> void {
+    try {
+        auto size = std::filesystem::file_size(file_path);
+        offset += size;
+    } catch (std::filesystem::filesystem_error const& e) {
+        throw OperationFailed(
+                ErrorCode_Failure,
+                __FILENAME__,
+                __LINE__,
+                fmt::format("Failed to get file size: {}", e.what())
+        );
+    }
+}
+
+auto
+get_file_infos(std::filesystem::path const& multi_file_archive_path, segment_id_t next_segment_id)
+        -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset = 0;
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
+        get_file_size_and_update_offset(multi_file_archive_path / static_archive_file_name, offset);
```

Comment:
Now looking at the code again, wouldn't it make more sense to:
- Let get_file_size_and_update_offset only return file_size, i.e. turn it to get_file_size. In this case, maybe we already have similar methods somewhere in shared utils. if not, maybe we can create one as share method instead of defining it only in this file.
- Let get_file_infos increment the offset.

Reply:
I feel it is ok. I remember Zhihao had a talk about error handling but I wasn't paying attention enough.

@LinZhihao-723, do you think we can directly throw the exception from std::filesystem ?